contrib/format-msgpack/src/main/java/org/apache/drill/exec/store/msgpack/MsgpackSchema.java

```diff
@@ -0,0 +1,114 @@
+
+  private DrillFileSystem fileSystem;
+
+  public MsgpackSchema(DrillFileSystem fileSystem) {
+    this.fileSystem = fileSystem;
+  }
+
+  public MaterializedField load(Path schemaLocation) throws AccessControlException, FileNotFoundException, IOException {
+    MaterializedField previousMapField = null;
+    if (schemaLocation != null && fileSystem.exists(schemaLocation)) {
+      try (FSDataInputStream in = fileSystem.open(schemaLocation)) {
+        String schemaData = IOUtils.toString(in);
+        Builder newBuilder = SerializedField.newBuilder();
+        try {
+          TextFormat.merge(schemaData, newBuilder);
+        } catch (ParseException e) {
+          throw new DrillRuntimeException("Failed to read schema file: " + schemaLocation, e);
+        }
+        SerializedField read = newBuilder.build();
+        previousMapField = MaterializedField.create(read);
+      }
+    }
+    return previousMapField;
+  }
+
+  public void save(MaterializedField mapField, Path schemaLocation) throws IOException {
+    try (FSDataOutputStream out = fileSystem.create(schemaLocation, true)) {
+      SerializedField serializedMapField = mapField.getSerializedField();
+      String data = TextFormat.printToString(serializedMapField);
+      IOUtils.write(data, out);
+    }
+  }
+
+  public MaterializedField merge(MaterializedField existingField, MaterializedField newField) {
+    Preconditions.checkArgument(existingField.getType().getMinorType() == MinorType.MAP,
+        "Field " + existingField + " is not a MAP type.");
+    Preconditions.checkArgument(existingField.hasSameTypeAndMode(newField),
+        "Field " + existingField + " and " + newField + " not same.");
+    MaterializedField merged = existingField.clone();
+    privateMerge(merged, newField);
+    return merged;
+  }
+
+  private void privateMerge(MaterializedField existingField, MaterializedField newField) {
+    Preconditions.checkArgument(existingField.getType().getMinorType() == MinorType.MAP,
+        "Field " + existingField + " is not a MAP type.");
+    for (MaterializedField newChild : newField.getChildren()) {
+      String newChildName = newChild.getName();
+      MaterializedField foundExistingChild = getFieldByName(newChildName, existingField);
+      if (foundExistingChild != null) {
+        if (foundExistingChild.hasSameTypeAndMode(newChild)) {
+          if (foundExistingChild.getType().getMinorType() == MinorType.MAP) {
+            privateMerge(foundExistingChild, newChild);
+          } // else we already have it
+        } else {
+          // error
+          throw new SchemaChangeRuntimeException("Not the same schema for " + foundExistingChild + " and " + newChild);
+        }
+      } else {
+        existingField.addChild(newChild.clone());
+      }
+    }
+  }
+
+  private MaterializedField getFieldByName(String newChildName, MaterializedField existingField) {
+    for (MaterializedField f : existingField.getChildren()) {

```

Comment:
A sequential search won't perform well if called once per row. Or, is this called only once per column?

As it turns out, the TupleSchema is designed with a hash map to speed up name access. Perhaps consider that mechanism if you need frequent access to these columns.